cluster-autoscaler/debuggingsnapshot/debugging_snapshot.go

```diff
@@ -77,6 +81,21 @@ func (s *DebuggingSnapshotImpl) SetUnscheduledPodsCanBeScheduled(podList []*v1.P
 	}
 }
 
+// SetTemplateNode is the setter for TemplateNode
+func (s *DebuggingSnapshotImpl) SetTemplateNode(templates []*framework.NodeInfo) {
+	if templates == nil {
+		return
+	}
+
+	var TemplateNodeInfo []*framework.NodeInfo
+	for _, template := range templates {
+		t := template.Clone()
```

Comment:
Why do you clone the nodeInfo here and create a new NodeInfo instead in SetNodeGroupInfo? It seems like the logic should be the same in both cases. Maybe consider extracting the copying logic to a helper function?

Reply:
we also clone in SetNodeGroupInfo along with transforming it into a new data type. The new data type debugging_snaphot.NodeInfo, contains *framework.Node and []*framework.Pod but it doesn't contain the "template", that is framwork.NodeInfo.

Reply:
Oh, I missed that it's using a locally defined "NodeInfo" struct. Can we rename that? Masking name like that makes it super confusing (as demonstrated by the fact that I didn't realize it was a different NodeInfo type).

Also - why the different representation? As far as I know framework.NodeInfo is a wrapper for a node + set of pods and everything else is just some precomputation to optimize scheduler functions. Is there any data we're missing by using the node+pods representation? If so, why is it ok to lose this data for nodes, but not ok to lose it for template nodes?

Reply:
Renamed.
The idea is to remove excess / duplicate information for "ClusterNodes". Since we already have the information in the node spec, we don't need another representation / copy in the Template. (At least for the fields I thought would of interest). This also has scalability implications, having duplicate template for every node in that NodeGroup.

Reply:
I'm confused. framework.NodeInfo is just a wrapper for node+pods used by scheduler. I don't think there is anything more to it (sure, there are other fields, but AFAIK those are just some precomputed values that are purely based on the node+pods). The template itself is just node + ds + static pods.
For convenience CA re-uses framework.NodeInfo to internally represent node+pods. We do it for both template nodes and actual nodes. As far as I know the only difference between 'cluster node' NodeInfo and 'template node' NodeInfo is that one represents a node that really exist and the other an in-memory fake node, but functionally they should be the same.

I don't think there is anything in NodeInfo that we need in snapshot other than node+pods. And if there were, I don't understand why we'd need it only for template nodes.

Reply:
Based on offline discussion, it is concluded that template node should also be of internal debugging data struct `ClusterNode`, as data needed would be available in the internal struct and there would be similarities of type usage.